crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,762 @@
+
+            let json_str = serde_json::to_string(&result).map_err(TransportErrorKind::custom)?;
+            let hash = req.params_hash()?;
+            let _ = cache.put(hash, json_str);
+
+            Ok(result)
+        }))
+    }
+}
+
+/// Internal type to handle different types of requests and generating their param hashes.
+struct RequestType<Params: RpcParam> {
+    method: &'static str,
+    params: Params,
+    block_id: Option<BlockId>,
+}
+
+impl<Params: RpcParam> RequestType<Params> {
+    const fn new(method: &'static str, params: Params) -> Self {
+        Self { method, params, block_id: None }
+    }
+
+    const fn with_block_id(mut self, block_id: BlockId) -> Self {
+        self.block_id = Some(block_id);
+        self
+    }
+
+    fn params_hash(&self) -> TransportResult<B256> {
+        // Merge the method + params and hash them.
+        let hash = serde_json::to_string(&self.params())
+            .map(|p| keccak256(format!("{}{}", self.method(), p).as_bytes()))
+            .map_err(RpcError::ser_err)?;
+
+        Ok(hash)
+    }
+
+    const fn method(&self) -> &'static str {
+        self.method
+    }
+
+    fn params(&self) -> Params {
+        self.params.clone()
+    }
+
+    /// Returns true if the BlockId has been set to a tag value such as "latest", "earliest", or
+    /// "pending".
+    const fn has_block_tag(&self) -> bool {
+        if let Some(block_id) = self.block_id {
+            match block_id {
+                BlockId::Hash(_) => return false,
+                BlockId::Number(BlockNumberOrTag::Number(_)) => return false,
+                _ => return true,
+            }
+        }
+        false
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct FsCacheEntry {
+    /// Hash of the request params
+    key: B256,
+    /// Serialized response to the request from which the hash was computed.
+    value: String,
+}
+
+/// Configuration for the cache layer.
+/// For future extensibility of the configurations.
+#[derive(Debug, Clone)]
+pub struct CacheConfig {
+    /// Maximum number of items to store in the cache.
+    pub max_items: u32,
+}
+
+/// Shareable cache.
+#[derive(Debug, Clone)]
+pub struct SharedCache {
+    inner: Arc<RwLock<LruMap<B256, String>>>,
+}
+
+impl SharedCache {
+    /// Instantiate a new shared cache.
+    pub fn new(max_items: u32) -> Self {
+        let inner = Arc::new(RwLock::new(LruMap::<B256, String>::new(ByLength::new(max_items))));
+        Self { inner }
+    }
+
+    /// Puts a value into the cache, and returns the old value if it existed.
+    pub fn put(&self, key: B256, value: String) -> TransportResult<bool> {
+        Ok(self.inner.write().insert(key, value))
+    }
+
+    /// Gets a value from the cache, if it exists.
+    pub fn get(&self, key: &B256) -> TransportResult<Option<String>> {
+        // Need to acquire a write guard to change the order of keys in LRU cache.
+        let val = self.inner.write().get(key).cloned();
```

Comment:
this only needs a read lock

Reply:
`.get` of LruMap is &mut self and requires a write lock as it promotes the key that was accessed if found